mypy/constraints.py

```diff
@@ -336,6 +338,38 @@ def visit_instance(self, template: Instance) -> List[Constraint]:
                     res.extend(infer_constraints(
                         template.args[j], mapped.args[j], neg_op(self.direction)))
                 return res
+            if (template.type.is_protocol and self.direction == SUPERTYPE_OF and
+                    # We avoid infinite recursion for structural subtypes by checking
+                    # whether this type already appeared in the inference chain.
+                    not any(is_same_type(template, t) for t in template.type.inferring) and
+                    mypy.subtypes.is_subtype(instance, erase_typevars(template))):
+                template.type.inferring.append(template)
+                for member in template.type.protocol_members:
+                    inst = mypy.subtypes.find_member(member, instance, original_actual)
+                    temp = mypy.subtypes.find_member(member, template, original_actual)
+                    assert inst is not None and temp is not None
+                    res.extend(infer_constraints(temp, inst, self.direction))
+                    if (mypy.subtypes.IS_SETTABLE in
+                            mypy.subtypes.get_member_flags(member, template.type)):
+                        # Settable members are invariant, add opposite constraints
+                        res.extend(infer_constraints(temp, inst, neg_op(self.direction)))
+                template.type.inferring.pop()
+                return res
+            elif (instance.type.is_protocol and self.direction == SUBTYPE_OF and
```

Comment:
This and the previous section of code look very similar. Can you merge them to avoid code duplication?